Visc_rem_[uv] multiplied momentum budget diagnostics

NoraLoose opened the pull request:
This PR implements modified diagnostics for the momentum budget.

### Modified momentum budget
 The modified momentum budget diagnostics 
* filter out the balance between `PFu` and `du_dt_visc` in vanished layers,
* leave the momentum budget as is in non-vanished layers.

In practice, this is done by modifying the original zonal momentum budget 
```
du_dt = CAu + PFu + u_BT_accel + diffu + du_dt_visc    (1)
```
to 
```
du_dt = [visc_rem_u * CAu] + [visc_rem_u * PFu] + [visc_rem_u * u_BT_accel] + [visc_rem_u * diffu] + [visc_rem_u * du_dt_visc + (1-visc_rem_u) * du_dt]   (2)
```
as per @Hallberg-NOAA's suggestion in https://github.com/ocean-eddy-cpt/WP1T4-2D-EKE-NW2/issues/8.

There are five terms in square brackets on the right hand side of (2). The first 4 are labelled as `CAu_visc_rem`, `PFu_visc_rem`, `u_BT_accel_visc_rem`, `diffu_visc_rem`. These are the 4 new implemented diagnostics. 

The 5th term, `[visc_rem_u * du_dt_visc + (1-visc_rem_u) * du_dt]`, has a slightly more complex form. To make our lives easier, I decided to drop the implementation for this 5th term, because it **can be diagnosed exactly as the residual** `du_dt - CAu_visc_rem - PFu_visc_rem - u_BT_accel_visc_rem - diffu_visc_rem` from existing diagnostics.

Everything that I am describing above is similarly done for the meridional momentum budget.

### Why are the new diagnostics useful?
The new diagnostics are helpful for my CPT work because I apply a spatial filter to each of the terms in the momentum and KE budget. This filter will spread high `PFu` / `du_dt_visc` values from vanished layers into the interior ocean, which contaminates my EKE budget. The problem is solved when the `PFu` / `du_dt_visc` balance is masked out in vanished layers from the get-go.

NoraLoose committed:
Implement visc remnant versions of PFu, CAu, u_accel_bt, diffu

NoraLoose committed:
Add d[uv]_dt_visc_rem diagnostics

NoraLoose committed:
Attempt to implement KE budget modified by visc rem

NoraLoose commented:
PS: The unit tests suggest that I made syntax errors in `MOM_diagnostics.F90`. I am confused - I am able to compile with intel on cheyenne just fine. :confused:

gustavo-marques commented on src/diagnostics/MOM_diagnostics.F90:
```diff
@@ -2300,8 +2438,11 @@ subroutine set_dependent_diagnostics(MIS, ADp, CDp, G, GV, CS)
       associated(CS%KE_adv) .or. associated(CS%KE_visc) .or. &
-      associated(CS%KE_horvisc) .or. associated(CS%KE_dia)) &
+      associated(CS%KE_horvisc) .or. associated(CS%KE_dia) .or. & 
+      associated(CS%PE_to_KE_visc_rem) .or. & associated(CS%KE_BT_visc_rem) .or. &
```
You have an extra "&" here.

NoraLoose replied:
Good catch. (This didn't fix the issue that I'm describing above, though.)

gustavo-marques commented:
I am happy to meet early next week to discuss the issues you are having. Let's coordinate that via Slack.

NoraLoose committed:
Fix syntax error

NoraLoose commented:
Thanks Gustavo! 
> Let's coordinate that via Slack.

Sounds good.

NoraLoose committed:
Fix some style errors

NoraLoose committed:
Fix more style errors

hmkhatri commented on src/diagnostics/MOM_diagnostics.F90:
```diff
@@ -1877,17 +1992,30 @@ subroutine MOM_diagnostics_init(MIS, ADp, CDp, Time, G, GV, US, param_file, diag
+  CS%id_KE_Coradv_visc_rem = register_diag_field('ocean_model', 'KE_Coradv_visc_rem', diag%axesTL, Time, &
+      'Kinetic Energy Source from Coriolis and Advection multiplied by viscous remnant fraction', &
+      'm3 s-3', conversion=GV%H_to_m*(US%L_T_to_m_s**2)*US%s_to_T)
+  if (CS%id_KE_Coradv_visc_rem>0) call safe_alloc_ptr(CS%KE_Coradv_visc_rem,isd,ied,jsd,jed,nz)
```
There might be an issue with character length (not sure as I don't see character length on github). The max character length should be less than 120 in each line. It might be the issue.

NoraLoose commented:
Thanks everyone for your help! With @gustavo-marques's help, I figured out what the issue was. I have now implemented both the momentum and KE budget diagnostics. Here are two summary plots that show the updated vs. original diagnostics in the lowermost layer of NW2 (5 day average):

### Momentum budget 

![mom_budget_u_visc_rem_layer14](https://user-images.githubusercontent.com/23617395/128075862-a0d1ed5a-306a-4304-826f-589fa15b0b38.png)

**Key point:** the `*_visc_rem` multiplied **momentum budget** diagnostics filter out the balancing terms of the pressure gradient force (`PFu`) and vertical viscosity (`du_dt_visc`) in *vanished layers*, as desired.

### KE budget

![KE_budget_visc_rem_layer14](https://user-images.githubusercontent.com/23617395/128076267-a1f478a4-0bda-4842-b37c-85bdaf51712a.png)

**Key point:** the updated KE budget diagnostics (designed to be consistent with the `visc_rem_[uv]` multiplied momentum budget diagnostics) are identical to the original KE diagnostics. In hindsight, this makes sense: one dots the momentum budget diagnostics with `uh` to compute the KE budget diagnostics. The differences between the old and new momentum budget diags are found **only in vanished layers**, so the KE budget does not feel it.

Since the new and original KE budget diagnostics are identical (up to numerical noise), I am therefore now considering to take the KE budget diagnostics out of this PR.

NoraLoose committed:
Remove added KE budget diagnostics

NoraLoose committed:
Fix small bug

NoraLoose committed:
Small style fix

NoraLoose commented:
To keep it simple -- and to avoid redundant diagnostics --, I am restricting this PR to `visc_rem_[uv]` multiplied **momentum budget** diagnostics **only**. The associated KE budget diagnostics that were originally part of this PR are now removed.

The code changes are ready for review now.  All tests are passing except for one regression test that complains about the fact that the diagnostics have been changed.

gustavo-marques commented on src/diagnostics/MOM_diagnostics.F90:
```diff
@@ -290,6 +290,12 @@ subroutine calculate_diagnostic_fields(u, v, h, uh, vh, tv, ADp, CDp, p_surf, &
 
     if (CS%id_dh_dt>0) call post_data(CS%id_dh_dt, CS%dh_dt, CS%diag, alt_h = diag_pre_sync%h_state)
+    do k=1,nz ; do j=js,je ; do I=is-1,ie
+        ADp%du_dt(I,j,k) = CS%du_dt(I,j,k)
```
Perhaps you forgot to delete this loop?
If not, since you are modifying the ADp structure, its intent should be inout:
```
  type(accel_diag_ptrs),   intent(inout)    :: ADp  !< structure with pointers to
                                                 !! accelerations in momentum equation.
```

Also, this loop should only be exercised if CS%id_hf_du_dt_2d > 0. Otherwise, there will be a memory error.

NoraLoose replied:
Good point about the `inout`! I need `ADp%du_dt` for the computation of `du_dt_visc_rem`, see here:

https://github.com/NoraLoose/MOM6/blob/86741cd1d3b049303db325d61089263eda90979b/src/parameterizations/vertical/MOM_vert_friction.F90#L535-L536

gustavo-marques replied:
Gotcha! Then, you should also allocate  `ADp%du_dt` when `CS%id_du_dt_visc_rem > 0`.

NoraLoose replied:
Do you mean that I should add something like:
```
if (CS%id_du_dt_visc_rem > 0) then
    if (.not.associated(CS%du_dt)) then
        call safe_alloc_ptr(CS%du_dt,IsdB,IedB,jsd,jed,nz)
        call register_time_deriv(lbound(MIS%u), MIS%u, CS%du_dt, CS)
    endif
endif
```
in `MOM_diagnostics_init`?

gustavo-marques replied:
That won't work. In MOM_diagnostics_init, CS is the diagnostics_CS type.

gustavo-marques commented on src/diagnostics/MOM_diagnostics.F90:
```diff
@@ -290,6 +290,12 @@ subroutine calculate_diagnostic_fields(u, v, h, uh, vh, tv, ADp, CDp, p_surf, &
+    do k=1,nz ; do j=js,je ; do I=is-1,ie
+        ADp%du_dt(I,j,k) = CS%du_dt(I,j,k)
+    enddo ; enddo ; enddo
+    do k=1,nz ; do J=js-1,je ; do i=is,ie
```
Same here, it seems this should be deleted since `id_hf_du_dt` is commented out in `MOM_diagnostics_init`. This look should only be exercised if `ADp%dv_dt` has been allocated.

NoraLoose replied:
Same here. I need `ADp%dv_dt` to compute `dv_dt_visc_rem` in the `MOM_vert_friction` module. What do you recommend to do about allocation of `ADp%du_dt` and `ADp%dv_dt`? These variables are defined to be part of the `ADp` structure here:

https://github.com/NoraLoose/MOM6/blob/86741cd1d3b049303db325d61089263eda90979b/src/core/MOM_variables.F90#L164-L165

gustavo-marques replied:
I suggest allocating them in [MOM_diagnostics_init](https://github.com/NoraLoose/MOM6/blob/dev/gfdl/src/diagnostics/MOM_diagnostics.F90#L1556). 

du_dt from [diagnostics_CS](https://github.com/NoraLoose/MOM6/blob/dev/gfdl/src/diagnostics/MOM_diagnostics.F90#L70-L71) is only allocated when id_du_dt > 0, see [here](https://github.com/NoraLoose/MOM6/blob/dev/gfdl/src/diagnostics/MOM_diagnostics.F90#L1747-L1750). It also needs to be allocated when id_hf_du_dt > 0. Same for dv_dt.

gustavo-marques commented:
@NoraLoose, thanks for this PR. I was able to compile the code but it fails in the NW2 test because of the issues pointed above. I think it will run just fine that after you address them.

NoraLoose committed:
Fix allocation problems

NoraLoose committed:
Remove du_dt_visc_rem diagnostic

NoraLoose commented:
Thanks @gustavo-marques for walking me through the allocation issue!

### Dropping of `d[uv]_dt_visc_rem`

To make our lives easier, I have dropped the more complex diagnostic `du_dt_visc_rem`. This diagnostic is more complex than the other new diagnostics because it requires not only `du_dt_visc` and `visc_rem_u`, but also `du_dt`. Having to know `du_dt` resulted in complications with allocation across different modules / routines.

The diagnostic `du_dt_visc_rem` can be dropped because it can actually (by construction) be **exactly** diagnosed from diagnostics that are already there:
```
du_dt_visc_rem = du_dt - CAu_visc_rem - PFu_visc_rem - u_BT_accel_visc_rem - diffu_visc_rem
```
I have updated the PR description above accordingly.

### Testing 

I have tested the remaining 8 momentum budget diagnostics (4 zonal, 4 meridional) as follows. Recall that
```
PFu_visc_rem = PFu * visc_rem_u,
```
i.e. `PFu_visc_rem` is simply the `visc_rem_u`-multiplied version of `PFu`. Same holds for `CAu_visc_rem`, `u_BT_accel_visc_rem`, and `diffu_visc_rem`. 
* I wrote out `PFu_visc_rem`, `PFu`, and `visc_rem_u` on a *timestep* basis, and 
* compared `PFu_visc_rem` (online diagnostic) vs. `PFu` * `visc_rem_u` (offline diagnostic). These two should give the exact same result.

Here, I am showing this comparison for an example time step and the lowermost layer.

![visc_rem_mom_budget_online_vs_offline_layer14 ](https://user-images.githubusercontent.com/23617395/128268270-6e6e91c7-0897-4396-8837-1dc90addf83a.png)

The residual in the last column is orders of magnitude smaller than the signal, which gives me confidence that the new diagnostics are implemented correctly. I found the same for other layers as well as the meridional budget.

NoraLoose committed:
Fix small inconsistency: CS%visc_rem_v --> CS%ADp%visc_rem_v

NoraLoose committed:
Don't need inout for ADp anymore

gustavo-marques commented:
Thanks for the changes @NoraLoose! It looks good to me now. I've verified that it does not change answers in one of the NW2 tests.

NoraLoose commented:
Awesome! Thanks for your help and for handling this PR so quickly, @gustavo-marques!

Hallberg-NOAA commented:
Very nicely done @NoraLoose, and thanks for your help and being so responsive, @gustavo-marques.

  I think that others will find these new diagnostics to be very valuable as well, and would like to see these changes merged into the shared main branch, hopefully sooner rather than later.  Do we have a strategy for propagating these changes to the shared main branch, either by a PR to dev/ncar or dev/gfdl?

gustavo-marques commented:
Hi @Hallberg-NOAA, yes I will send a PR from dev/cpt to dev/gfdl to propagate these changes to the main branch.
